File: lv-py/src/lv_py/transformers/inputs.py
"""Input plugin transformers (Logstash inputs → Vector sources)."""

from lv_py.models import ComponentType
from lv_py.models.logstash_config import LogstashPlugin
from lv_py.models.vector_config import VectorComponent
from lv_py.transformers.base import BaseTransformer


class FileInputTransformer(BaseTransformer):
    """Transform Logstash file input to Vector file source."""

    def supports(self, plugin_name: str) -> bool:
        """Check if this transformer supports the given plugin."""
        return plugin_name == "file"

    def transform(self, plugin: LogstashPlugin) -> VectorComponent:
        """
        Transform Logstash file input to Vector file source.

        Logstash file input config:
            path => "/var/log/*.log"
            start_position => "beginning"
            sincedb_path => "/dev/null"

        Vector file source config:
            type = "file"
            include = ["/var/log/*.log"]
            read_from = "beginning"
        """
        config = plugin.config

        # Build Vector config
        vector_config = {}

        # Map path to include (Vector uses array)
        if "path" in config:
            path_value = config["path"]
            # If it's a string, convert to list
            if isinstance(path_value, str):
                vector_config["include"] = [path_value]
            elif isinstance(path_value, list):
                vector_config["include"] = path_value
        else:
            vector_config["include"] = ["/var/log/*.log"]  # Default

        # Map start_position to read_from
        start_position = config.get("start_position", "end")
        if start_position == "beginning":
            vector_config["read_from"] = "beginning"
        else:
            vector_config["read_from"] = "end"

        # Optionally map other fields
        if "tags" in config:
            # Add as metadata in comments since Vector handles this differently
            pass

        return VectorComponent(
            component_type=ComponentType.SOURCE,
            component_kind="file",
            config=vector_config,
            inputs=[],
            comments=[],
        )


class BeatsInputTransformer(BaseTransformer):
    """Transform Logstash beats input to Vector HTTP source."""

    def supports(self, plugin_name: str) -> bool:
        """Check if this transformer supports the given plugin."""
        return plugin_name == "beats"

    def transform(self, plugin: LogstashPlugin) -> VectorComponent:
        """
        Transform Logstash beats input to Vector HTTP or socket source.

        Logstash beats input:
            port => 5044
            host => "0.0.0.0"

        Vector equivalent (using socket source for Lumberjack protocol):
            type = "socket"
            address = "0.0.0.0:5044"
            mode = "tcp"
        """
        config = plugin.config

        # Build Vector config
        port = config.get("port", 5044)
        host = config.get("host", "0.0.0.0")
        address = f"{host}:{port}"

        vector_config = {
            "address": address,
            "mode": "tcp",
        }

        comments = [
            "NOTE: Beats input migrated to TCP socket",
            "For full Beats protocol support, consider using Filebeat → Vector directly",
        ]

        return VectorComponent(
            component_type=ComponentType.SOURCE,
            component_kind="socket",
            config=vector_config,
            inputs=[],
            comments=comments,
        )

File: lv-go/internal/transformer/inputs.go
package transformer

import (
	"fmt"

	"github.com/yourusername/lv-go/internal/models"
)

// FileInputTransformer transforms Logstash file input to Vector file source
type FileInputTransformer struct{}

func (t *FileInputTransformer) Supports(pluginName string) bool {
	return pluginName == "file"
}

func (t *FileInputTransformer) Transform(plugin models.LogstashPlugin) (*models.VectorComponent, error) {
	config := plugin.Config
	vectorConfig := make(map[string]interface{})

	// Map path to include (Vector uses array)
	if pathValue, ok := config["path"]; ok {
		switch v := pathValue.(type) {
		case string:
			vectorConfig["include"] = []string{v}
		case []interface{}:
			paths := make([]string, len(v))
			for i, p := range v {
				paths[i] = fmt.Sprint(p)
			}
			vectorConfig["include"] = paths
		case []string:
			vectorConfig["include"] = v
		}
	} else {
		vectorConfig["include"] = []string{"/var/log/*.log"} // Default
	}

	// Map start_position to read_from
	startPosition, _ := config["start_position"].(string)
	if startPosition == "" {
		startPosition = "end"
	}

	if startPosition == "beginning" {
		vectorConfig["read_from"] = "beginning"
	} else {
		vectorConfig["read_from"] = "end"
	}

	return &models.VectorComponent{
		ComponentType: models.ComponentTypeSource,
		ComponentKind: "file",
		Config:        vectorConfig,
		Inputs:        []string{},
		Comments:      []string{},
	}, nil
}

// BeatsInputTransformer transforms Logstash beats input to Vector socket source
type BeatsInputTransformer struct{}

func (t *BeatsInputTransformer) Supports(pluginName string) bool {
	return pluginName == "beats"
}

func (t *BeatsInputTransformer) Transform(plugin models.LogstashPlugin) (*models.VectorComponent, error) {
	config := plugin.Config

	// Get port and host
	port := 5044
	if p, ok := config["port"].(int); ok {
		port = p
	}

	host := "0.0.0.0"
	if h, ok := config["host"].(string); ok {
		host = h
	}

	address := fmt.Sprintf("%s:%d", host, port)

	vectorConfig := map[string]interface{}{
		"address": address,
		"mode":    "tcp",
	}

	comments := []string{
		"NOTE: Beats input migrated to TCP socket",
		"For full Beats protocol support, consider using Filebeat → Vector directly",
	}

	return &models.VectorComponent{
		ComponentType: models.ComponentTypeSource,
		ComponentKind: "socket",
		Config:        vectorConfig,
		Inputs:        []string{},
		Comments:      comments,
	}, nil
}
